### PR TITLE
feat: Add option to add tolerance to improvement in metric evolution

### PR DESCRIPTION
**Pull request by JoanFM**
feat: Implement feature requested in https://github.com/Azure/mmlspark/issues/713

_"""Even by using a validation set and enabling early stopping, the training of LightGBMRanker often reaches the maximum number of iterations. However, the metric computed using the validation set often differs by only a small margin between two iterations, something after the 4 decimal point.

Wouldn't be convenient to be able to set a tolerance t, such as that training stops after x rounds if the evaluation metric has improved less than t in x iterations? This should speed up the training phase, no?"""_

### CODE REVIEW

**Comment by imatiach-msft** [comment]
@JoanFM the PR looks great, could you please rebase against latest master and fix the merge conflicts?

**Comment by imatiach-msft** [comment]
Looks like some style checks failed - are you able to see them in the build?

[error] /home/vsts/work/1/s/src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala:64: File line length exceeds 120 characters

[error] /home/vsts/work/1/s/src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala:88: File line length exceeds 120 characters

[error] /home/vsts/work/1/s/src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala:64:123: Whitespace at end of line

[error] /home/vsts/work/1/s/src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala:88:122: Whitespace at end of line

File: src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala [hunk 1]
```diff
@@ -60,8 +61,8 @@ abstract class TrainParams extends Serializable {
 case class ClassifierTrainParams(parallelism: String, topK: Int, numIterations: Int, learningRate: Double,
                                  numLeaves: Int, maxBin: Int, binSampleCount: Int,
                                  baggingFraction: Double, posBaggingFraction: Double, negBaggingFraction: Double,
-                                 baggingFreq: Int, baggingSeed: Int, earlyStoppingRound: Int, featureFraction: Double,
-                                 maxDepth: Int, minSumHessianInLeaf: Double,
+                                 baggingFreq: Int, baggingSeed: Int, earlyStoppingRound: Int, improvementTolerance: Double,
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala]: looks like there is whitespace at end of line here that needs to removed.  Also line is too long.  You can just add newline after the earlyStoppingRound parameter.

Comment (JoanFM) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/TrainParams.scala]: Done, sorry I fixed merge issues directly on github, so I did not get the highlights of the issues

**Comment by imatiach-msft** [comment]
hmm seeing error:

[info] - Serialization Fuzzing *** FAILED ***

[info]   scala.NotImplementedError: The default jsonEncode only supports string, vector and matrix. org.apache.spark.ml.param.Param must override jsonEncode for java.lang.Double.

[info]   at org.apache.spark.ml.param.Param.jsonEncode(params.scala:100)

[info]   at org.apache.spark.ml.util.DefaultParamsWriter$$anonfun$6.apply(ReadWrite.scala:463)

[info]   at org.apache.spark.ml.util.DefaultParamsWriter$$anonfun$6.apply(ReadWrite.scala:462)

[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)

[info]   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)

[info]   at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)

[info]   at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)

[info]   at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)

[info]   at scala.collection.AbstractTraversable.map(Traversable.scala:104)

[info]   at org.apache.spark.ml.util.DefaultParamsWriter$.getMetadataToSave(ReadWrite.scala:462)

[info]   ...

[info]   + Creating a spark session for suite VerifyLightGBMClassifier

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala [hunk 2]
```diff
@@ -55,6 +55,23 @@ trait LightGBMExecutionParams extends Wrappable {
...
+/** Defines common parameters across all LightGBM learners related to learning score evolution.
+  */
+trait LightGBMLearnerParams extends Wrappable {
+  val earlyStoppingRound = new IntParam(this, "earlyStoppingRound", "Early stopping round")
+  setDefault(earlyStoppingRound -> 0)
+
+  def getEarlyStoppingRound: Int = $(earlyStoppingRound)
+  def setEarlyStoppingRound(value: Int): this.type = set(earlyStoppingRound, value)
+
+  val improvementTolerance = new Param[Double](this, "improvementTolerance",
```
Comment (imatiach-msft) [inline src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala]: change to DoubleParam

Comment (imatiach-msft) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala]: Param[Double] -> DoubleParam

**Comment by imatiach-msft** [comment]
I think changing Param[Double] to DoubleParam should fix the serialization issue

File: src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala (hunk 2 above)
Comment (JoanFM) [reply src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMParams.scala]: Done

**Pull request merged**